### DISCUSSION TITLE
can i use a outside variable to join messages two-by-two

### DISCUSSION THREAD

**Discussion by Matheus8174** [quirrel-dev/quirrel#1158] [Q&A]
i have no experience working with queues specially in NextJs so a question that is in my mind it's a good approach have a variable to join messsages in my case i wanna have it two by two, like in the code below

``` typescript
type MessageShape = {
    data: string
}

const messages: MessageShape[] = [];

export const chatQueue = Queue<MessageShape>(
  "api/queues/chat",
  async job => {
    messages.push(job);

    const output = messages.map((element, index) => {
      if (index % 2 === 0) {
        if (index + 1 < messages.length)
          return [element, messages[index + 1]];

        return [element];
      }
    });

    console.log(output);
  }
)
```

as you can see the messages variable is outside and will store all income messages, is it a good approach or there is a better way?

**Comment by Skn0tt** [comment]
When running your API Routes via Serverless, different requests will be handled by different Lambda instances. They don't share  their variables, so this won't work as you think it will - each job is executed on a different handler, so they all start fresh with an empty `messages` array.

I'd recommend you keep your API stateless, and store this data in an external database - for example Postgres or Redis.

**Comment by Matheus8174** [comment]
I'll use the same redis instance that is used by Quirrel to store it, any problem?. Just one more question, i can't use the enqueue method at the front-end side because the next can only access env variables by typing NEXT_PUBLIC at the variable's name beginning then i put the enqueue method in an api route. I am using Next13 btw.

**Reply by Skn0tt** [reply]
> I'll use the same Redis instance that is used by Quirrel to store it, any problem?
That should be fine, as long as you make sure you don't override any Quirrel keys. You can use https://redis.io/commands/select/ for that, in IORedis for example you can declare this via https://redis.github.io/ioredis/interfaces/CommonRedisOptions.html#db.

> i can't use the enqueue method at the front-end side

Yes, that's by design. The client-side of a web application isn't trustworthy - any user could open up the developer tools, and do things they're not supposed to do. If it would be possible to enqueue via the client-side, then an attacker could enqueue jobs breaking your application. You don't want that, so it's not possible. As you said, you should put the enqueue method into an API Route / Route Handler / `getServerSideProps` or similar - as long as it runs on the server-side and you perform the necessary authentication & authorization, you're safe.
